### PR TITLE
[multistage][hotfix] shuffle rewrite

### PR DESCRIPTION
**Pull request by walterddr**
shuffle exchange rewrite rules are not carrying partition correctly.

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java [hunk 1]
```diff
@@ -54,7 +55,7 @@ public class ShuffleRewriteVisitor implements StageNodeVisitor<Set<Integer>, Voi
    * @param root the root node of the tree to rewrite
    */
   public static void optimizeShuffles(StageNode root) {
-    root.visit(new ShuffleRewriteVisitor(), null);
+     root.visit(new ShuffleRewriteVisitor(), null);
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:58]:
```suggestion
    root.visit(new ShuffleRewriteVisitor(), null);
```

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java [hunk 2]
```diff
@@ -150,19 +147,11 @@ public Set<Integer> visitMailboxSend(MailboxSendNode node, Void context) {
...
-    // all inputs carry over if they're still in the projection result
-    Set<Integer> partitionKeys = new HashSet<>();
-    for (int i = 0; i < node.getProjects().size(); i++) {
-      RexExpression rex = node.getProjects().get(i);
-      if (rex instanceof RexExpression.InputRef) {
-        if (oldPartitionKeys.contains(((RexExpression.InputRef) rex).getIndex())) {
-          partitionKeys.add(i);
-        }
-      }
+    if (rexExpressionListContainsAllPartitionKey(node.getProjects(), oldPartitionKeys)) {
```
Comment (agavra) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:150]: why is this necessary for project? projects shouldn't change partitioning, correct?

Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:150]: project can make a partition become unpartitioned. for example
```
- Aggregation [ GroupSet = (0, 1, 2), AggCall = (...) ]
  - Project [ SelectList = [inputRef(0), inputRef(1), func(xx)], PartitionKeys = [] ]
    - MailboxReceived [ PartitionKeys = [inputRef(0), inputRef(1), inputRef(2)] ]
```
we need to remove the PartitionKeys (no longer inputRef [0, 1, 2]) because even the group set is now (0, 1, 2) it is no longer the same 0, 1, 2 it was referring to

Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:150]: actually there's another bug. fixing

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java [hunk 3]
```diff
@@ -189,4 +173,27 @@ private static boolean canSkipShuffle(Set<Integer> partitionKeys, KeySelector<Ob
...
     return false;
   }
+
+  private static Set<Integer> deriveNewPartitionKeysFromRexExpressions(List<RexExpression> rexExpressionList,
+      Set<Integer> oldPartitionKeys) {
+    Map<Integer, Integer> partitionKeyMap = new HashMap<>();
+    for (int i = 0; i < rexExpressionList.size(); i++) {
+      RexExpression rex = rexExpressionList.get(i);
+      if (rex instanceof RexExpression.InputRef) {
+        // put the old-index to new-index mapping
```
Comment (ankitsultana) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:183]: I have arrived to something similar in this: https://github.com/apache/pinot/pull/9873/files

That PR is still being tested.. we can sync up later for it.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java [hunk 4]
```diff
@@ -150,25 +145,14 @@ public Set<Integer> visitMailboxSend(MailboxSendNode node, Void context) {
...
-
-    return partitionKeys;
+    return deriveNewPartitionKeysFromRexExpressions(node.getProjects(), oldPartitionKeys);
   }
 
   @Override
   public Set<Integer> visitSort(SortNode node, Void context) {
-    // sort doesn't change the partition keys
-    return node.getInputs().get(0).visit(this, context);
+    // with sort shuffling reorder could change, so we can't directly return the old partition keys.
```
Comment (ankitsultana) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:153]: Can you share how this happens? My understanding is that sort happens within a stage and doesn't rename/remove any columns so we should be able to carry-over the keys.

**Review by ankitsultana — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java (hunk 4 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:153]: sortNode result record type will put collation columns in front of the rest of the selections. which essentially act as a "project". so it should follow similar rules to reorder the partitionkey into its new inputRef indices. 

since this is a hotfix PR I am not including this in it

Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:153]: CC @agavra ^

**Review by siddharthteotia — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java (hunk 4 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java]: synced offline. turns out if there's any re-ordering a project will always be added. so no need for handling this

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java [hunk 5]
```diff
@@ -79,13 +79,39 @@ public void testQueryWithException(String query, String exceptionSnippet) {
...
+    query = "SELECT a.col1, AVG(b.col3) FROM a JOIN b ON a.col1 = b.col2 AND a.col2 = b.col2"
+        + " WHERE a.col3 >= 0 AND a.col2 = 'a' AND b.col3 < 0 GROUP BY a.col1";
+    queryPlan = _queryEnvironment.planQuery(query);
+    assertGroupBySingletonAfterJoin(queryPlan, false);
+
+    // mismatched group key should not rewrite into singleton
+    query = "SELECT a.col3, a.col1, AVG(b.col3) FROM a JOIN b ON a.col1 = b.col2 AND a.col2 = b.col2"
+        + " WHERE a.col3 >= 0 AND a.col2 = 'a' AND b.col3 < 0 GROUP BY a.col1, a.col3";
+    queryPlan = _queryEnvironment.planQuery(query);
+    assertGroupBySingletonAfterJoin(queryPlan, false);
```
Comment (61yao) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java:111]: Can we add a test case for expanding group key? 

 query = "SELECT a.col1, a.col2, AVG(b.col3) FROM a JOIN b ON a.col1 = b.col2 GROUP BY a.col1, a.col2";

**Comment by 61yao** [comment]
Can we also add query option to disable shuffle if needed. So we can know 1) the performance gain from this 2) remove the shuffle when we see an issue

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java [hunk 6]
```diff
@@ -189,4 +172,27 @@ private static boolean canSkipShuffle(Set<Integer> partitionKeys, KeySelector<Ob
...
+  private static Set<Integer> deriveNewPartitionKeysFromRexExpressions(List<RexExpression> rexExpressionList,
+      Set<Integer> oldPartitionKeys) {
+    Map<Integer, Integer> partitionKeyMap = new HashMap<>();
+    for (int i = 0; i < rexExpressionList.size(); i++) {
+      RexExpression rex = rexExpressionList.get(i);
+      if (rex instanceof RexExpression.InputRef) {
+        // put the old-index to new-index mapping
+        // TODO: it doesn't handle duplicate references. e.g. if the same old partition key is referred twice. it will
+        // only keep the second one. (see JOIN handling on left/right as another example)
+        partitionKeyMap.put(((RexExpression.InputRef) rex).getIndex(), i);
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ShuffleRewriteVisitor.java:185]: technically this is not a "bug" b/c the partitionkey generated this rewrite is in fact "CORRECT".

both the join situation and the project with multiple select are considered as "equivalent" partition and affects the "canSkipShuffle" conditions. these also include:

1. transformed version instead of just inputRef, such as ` 1 - partitionKey1`
2. mixed of transformed with all partitionKeys included and guaranteed to be unique, such as `makeListUDF(partitionKey1, partitionKey2)`
3. ...

There can be many more, we can iteratively add support for these semantics

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java (hunk 5 above)
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java:111]: I am not sure I understand the example, this exact query (without the WHERE clause is already included)

**Comment by walterddr** [comment]
@61yao merging this first to unblock some other PRs. please let me know if anything else requires follow up.

> Can we also add query option to disable shuffle if needed. So we can know 1) the performance gain from this 2) remove the shuffle when we see an issue

yes this can be done in follow up PRs.
